src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php

```diff
@@ -517,7 +542,9 @@ private function doRead($sessionId)
                 return is_resource($sessionRows[0][0]) ? stream_get_contents($sessionRows[0][0]) : $sessionRows[0][0];
             }
 
-            if (self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+            if (!ini_get('session.use_strict_mode') && self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+                // In strict mode, session fixation is not possible: new sessions always start with a unique
+                // random id, so that concurrency is not possible and this code path can be skipped.
```

Comment:
I don't agree. If the session just expired or was deleted manually, there can still be race conditions.

Reply:
Can you describe the scenario?
To me, when the session expired, or when the session was deleted in another place,
*in strict mode only*, php will generate a new id, so that there is no race possible.

Reply:
I wonder how php knows if the session is new? so when you do session_start() with an old id which expired, the save handler is executed which returns an empty string. And then php finishes the save handler by calling close() etc and then generates a new session id? Does it call the save handler open, read etc again?

Reply:
Thanks to the new interface, PHP calls "validateId", which return false if the handler's "read" returns the empty string.
Then, PHP calls "read()" directly, but now with a new random id.

Reply:
This is important also for the advisory locks which might otherwise not be released at all if it just calls save() with a new id.

Reply:
Anything to change in your opinion? The advisory lock looks also good to me as is.

Reply:
> Thanks to the new interface, PHP calls "validateId", which returns false if the handler's "read" returns the empty string. Then, PHP calls "read()" directly, but now with a new random id.

That does not make sense to me. When I start a new session without having a cookie, PHP generates a new session id which it then calls `validateId` with. As we have no data, it returns false. So php generates a new session id again, and calls `read` which again return no data. So to start a new session, it tries to read two different session ids twice?

Reply:
Validation happens only if there is a cookie of course.

Reply:
That does not seem to be true. The storage.expected tests shows what I was questioning

```
validateId
read
doRead: 
read
doRead: 
```

and there does not seem to be a cookie set before.

Reply:
The issue with going with only a proxy is that it will prevent skipping the lock in PdoSessionHandler, because we couldn't know if we're having the required logic or not.

Reply:
Oh, and implementating updateTimestamp per-handler would have no contract to be based on (and adding one would be worse.)

Reply:
I think the validation is only about checking if the session id has invalid characters in it according to the patch for strict sessions https://gist.github.com/yohgaki/1379668#file-php-trunk-strict-session-patch-L55

Thus I don't think validation should actually read the data.

Reply:
What I wonder is, if php implements a default validation of the id, e.g. by checking the length of the session id against `session.sid_length` php ini. Something like that seems more the correct approach to me.

Reply:
Reading https://wiki.php.net/rfc/strict_sessions

> This patch adds validate_id() function to ps_modules (session save handler) that can check if the session ID is already initialized one or not. Bundled ps_modules are now have validate function and will never accept uninitialized session ID

it suggests validate id is really about checking whether the session is initialized. Then the current solution is fine. I'd suggest to prevent new session from being read twice, to do

```
if ($prefetchId === $sessionId || '' === $prefetchData) {
    return $prefetchData;
}
```

Reply:
Back to the original question. With the validateId and strict session in place, race conditions on new session ids cannot happen indeed. So it's fine to skip inserting fake rows.

Reply:
> if ($prefetchId === $sessionId || '' === $prefetchData) {

good idea, done